### PR TITLE
Add a `CapsuleTest` trait

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

The purpose of this trait is to allow tests to have a callback when they are finished.

The trait:

```rust
pub trait CapsuleTestClient {
    fn done(&'static self, result: Result<(), CapsuleTestErr>);
}

/// Identify a test as a capsule test. This is only used for setting the client
/// for test complete callbacks.
pub trait CapsuleTest {
    fn set_client(&self, client: &'static dyn CapsuleTestClient);
}
```

I updated a couple tests (the tests that the nrf52840dk supports) with this functionality.

With this trait, a board can iterate through a list of tests one at a time.

### Testing Strategy

This pull request was tested by testing the nrf52840dk tests.

### TODO or Help Wanted

n/a

### Documentation Updated

I will update https://book.tockos.org/development/tests if this PR is acceptable.

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/core/src/test/capsule_test.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Interface for running capsule tests.
+//!
+//! As Tock capsules are asynchronous, it is difficult for a test runner to
+//! determine when a test when a test has finished. This interface provides a
+//! `done()` callback to notify when the test is done.
```
Comment (alistair23) [inline capsules/core/src/test/capsule_test.rs:9]: This doesn't seem right. The tests should be testing the same flow that the kernel and apps will use.

If it is too difficult to determine if a operation has finished then the HIL should be improved to make it easier. Adding special callbacks that only tests will use will just lead to testing different use cases then we are actually running. Which then isn't really testing anything useful

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: capsules/core/src/test/capsule_test.rs (hunk 1 above)
Comment (bradjc) [reply capsules/core/src/test/capsule_test.rs:9]: I'm not sure I follow. I updated the tests in with this PR, how would the test runner know when the test is finished? A test might cover multiple HILs, so the test runner cannot (always) use a HIL interface. Also, a test might use one asynchronous operation, or a thousand. Only the test can know when it is done.

Comment (alistair23) [reply capsules/core/src/test/capsule_test.rs:9]: Ah. I misunderstood. I thought you were wanting to add this to any capsule, not just test capsules

File: capsules/core/src/test/capsule_test.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Interface for running capsule tests.
+//!
+//! As Tock capsules are asynchronous, it is difficult for a test runner to
+//! determine when a test when a test has finished. This interface provides a
```
Comment (alistair23) [inline capsules/core/src/test/capsule_test.rs:8]: determine when a test has finished

File: capsules/core/src/test/capsule_test.rs (hunk 1 above)
Comment (alistair23) [inline capsules/core/src/test/capsule_test.rs:9]: `done()` callback to be used by testing capsule to notify when the test is done.

File: capsules/core/src/test/capsule_test.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//!         });
+//!     }
+//! }
+//! ```
+
+/// Client for receiving test done events.
+pub trait CapsuleTestClient {
+    /// Called when the test is finished. If test was successful, `result` is
+    /// `Ok(())`. If test failed, `result` is `Err(())`.
+    fn done(&'static self, result: Result<(), ()>);
```
Comment (alistair23) [inline capsules/core/src/test/capsule_test.rs:46]: Should we propagate the ErrorCode as well?

Comment (bradjc) [reply capsules/core/src/test/capsule_test.rs:46]: Is there an `ErrorCode` to return? I guess there would be if the test fails due to an error being returned in the underlying driver. But if the test finishes, but with the wrong answer?

I could see adding a new error type that is relevant to testing.

Comment (alistair23) [reply capsules/core/src/test/capsule_test.rs:46]: Exactly, there will be an `ErrorCode` if some underlying hardware/HIL/capsule returns an error.

The wrong answer is a different problem. We could either reuse an existing `ErrorCode` or create a special test one

**Review by alistair23 — APPROVED** [review]

File: capsules/extra/src/test/siphash24.rs [hunk 4]
```diff
@@ -56,10 +59,33 @@ impl Client<8> for TestSipHash24 {
...
-        debug!("expected result: {:?}", self.correct_hash.take().unwrap());
+        let correct = self.correct_hash.take().unwrap();
+
+        let mut matches = true;
+        for i in 0..8 {
+            if correct[i] != digest[i] {
+                matches = false;
+                kernel::debug!("TestSipHash24: incorrect hash output!");
+            }
+        }
```
Comment (hudson-ayers) [inline capsules/extra/src/test/siphash24.rs:70]:
```suggestion
        let matches = correct != digest;
        if !matches {
            kernel::debug!("TestSipHash24: incorrect hash output!");
        }
```
You can just compare equal length arrays directly!

**Review by hudson-ayers — APPROVED** [review]

**Review by alevy — DISMISSED** [review]

**Review by alevy — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]
This makes sense to me.

> I will update https://book.tockos.org/development/tests if this PR is acceptable.

I'd appreciate that. For me it took a while to realize that this is taking the subsystem-specific callback types / events and maps them onto a single callback that is generic across all tests, but pluggable from within the board's test infrastructure.

**Pull request merged**